[jvm-mixin] Error building

Hello, i try to build the jvm mixin, but I've got this error:

```
$  jb install
GET https://github.com/grafana/jsonnet-libs/archive/09ca0a6735fa87a36cccd2e74fb9734eb0a88ac4.tar.gz 200
GET https://github.com/grafana/jsonnet-libs/archive/09ca0a6735fa87a36cccd2e74fb9734eb0a88ac4.tar.gz 200
GET https://github.com/grafana/grafonnet/archive/d20e609202733790caf5b554c9945d049f243ae3.tar.gz 200
GET https://github.com/yugui/jsonnetunit/archive/6927c58cae7624a00f368b977ccc477d4f74071f.tar.gz 200
GET https://github.com/jsonnet-libs/docsonnet/archive/6ac6c69685b8c29c54515448eaca583da2d88150.tar.gz 200
GET https://github.com/jsonnet-libs/xtd/archive/1199b50e9d2ff53d4bb5fb2304ad1fb69d38e609.tar.gz 200
GET https://github.com/grafana/jsonnet-libs/archive/09ca0a6735fa87a36cccd2e74fb9734eb0a88ac4.tar.gz 200

$ make build
mixtool generate all mixin.libsonnet
2025/02/19 08:34:07 RUNTIME ERROR: Missing argument: nameShort
        /Users/nicolas.lamirault/Projects/monitoring-mixins/TMP/jsonnet-libs/jvm-mixin/vendor/common-lib/common/signal/signal.libsonnet:(153:14)-(157:15)       object <anonymous>
        /Users/nicolas.lamirault/Projects/monitoring-mixins/TMP/jsonnet-libs/jvm-mixin/vendor/process-observ-lib/panels.libsonnet:14:9-37
        /Users/nicolas.lamirault/Projects/monitoring-mixins/TMP/jsonnet-libs/jvm-mixin/vendor/process-observ-lib/rows.libsonnet:13:11-29        thunk from <thunk from <object <anonymous>>>
        /Users/nicolas.lamirault/Projects/monitoring-mixins/TMP/jsonnet-libs/jvm-mixin/vendor/github.com/grafana/grafonnet/gen/grafonnet-v11.0.0/custom/util/grid.libsonnet:150:57-62
        /Users/nicolas.lamirault/Projects/monitoring-mixins/TMP/jsonnet-libs/jvm-mixin/vendor/github.com/grafana/grafonnet/gen/grafonnet-v11.0.0/custom/util/grid.libsonnet:95:31-36    thunk from <thunk <gridPos> from <function <calculateGridPosForPanel>>>
        <std>:1536:24-25        thunk from <function <anonymous>>
        <std>:1536:8-41 function <anonymous>
        /Users/nicolas.lamirault/Projects/monitoring-mixins/TMP/jsonnet-libs/jvm-mixin/vendor/github.com/grafana/grafonnet/gen/grafonnet-v11.0.0/custom/util/grid.libsonnet:95:23-52    thunk <gridPos> from <function <calculateGridPosForPanel>>
        /Users/nicolas.lamirault/Projects/monitoring-mixins/TMP/jsonnet-libs/jvm-mixin/vendor/github.com/grafana/grafonnet/gen/grafonnet-v11.0.0/custom/util/grid.libsonnet:96:29-36    thunk from <thunk <width> from <function <calculateGridPosForPanel>>>
        <std>:1536:24-25        thunk from <function <anonymous>>
        ...
        /Users/nicolas.lamirault/Projects/monitoring-mixins/TMP/jsonnet-libs/jvm-mixin/vendor/github.com/grafana/grafonnet/gen/grafonnet-v11.0.0/custom/util/panel.libsonnet:(216:5)-(228:6)    function <anonymous>
        /Users/nicolas.lamirault/Projects/monitoring-mixins/TMP/jsonnet-libs/jvm-mixin/vendor/jvm-observ-lib/dashboards.libsonnet:(11:11)-(37:12)       thunk from <object <anonymous>>
        /Users/nicolas.lamirault/Projects/monitoring-mixins/TMP/jsonnet-libs/jvm-mixin/vendor/github.com/grafana/grafonnet/gen/grafonnet-v11.0.0/custom/dashboard.libsonnet:29:30-36    thunk from <object <anonymous>>
        <std>:32:25-26  thunk from <function <anonymous>>
        <std>:32:16-27  function <anonymous>
        /Users/nicolas.lamirault/Projects/monitoring-mixins/TMP/jsonnet-libs/jvm-mixin/vendor/github.com/grafana/grafonnet/gen/grafonnet-v11.0.0/custom/dashboard.libsonnet:29:18-37    object <anonymous>
        /Users/nicolas.lamirault/Projects/monitoring-mixins/TMP/jsonnet-libs/jvm-mixin/vendor/github.com/grafana/grafonnet/gen/grafonnet-v11.0.0/custom/dashboard.libsonnet:33:12-24    object <anonymous>
        Field "panels"
        Field "jvm-dashboard.json"
        During manifestation
make: *** [build] Error 1
```

--- a/common-lib/CHANGELOG.md
+++ b/common-lib/CHANGELOG.md
@@ -1,7 +1,11 @@
+# 0.3.1
+- [Signal] Fix optional signals (type=stub).
+- [Signal] Add optional signals documentation.
+
 # 0.3.0
 
-- [Signal] add new signal `log`.
-- [Signal] add enableLokiLogs=true|false to signals init.
+- [Signal] Add new signal `log`.
+- [Signal] Add enableLokiLogs=true|false to signals init.
 - [Signal] `withExprWrappersMixin(offset)` - wrap signal expression into additional function on top of existing wrappers.
 
 

--- a/common-lib/common/signal/README.md
+++ b/common-lib/common/signal/README.md
@@ -91,6 +91,7 @@ Signal's level:
 |name|Signal's name. Used to populate panel's titles. |*|CPU utilization|-|
 |nameShort|Signal's short name. Used to populate panel's legends and column names. |*|CPU|-|
 |type|Signal's type. Depending on the type, some opinionated autotransformations would happen with queries, units. |gauge,counter,histogram,info,raw|gauge|-|
+|optional| Set this signal optional.| true,false | false | false|
 |unit| Signal's units. |*|bytes|``|
 |description| Signal's description. Used to populate panel's description. |*|CPU usage time in percent.|``|
 |sourceMaps[].expr| Signal's BASE expression in simplest form. Simplified jsonnet templating is supported (see below). Depending on signal's type(not `raw`) could autotransform to different form. |*|network_bytes_received_total{%(queriesSelector)s}|-|
@@ -117,6 +118,9 @@ The following is supported in expressions and legends:
 - `%(interval)s` - expands to `interval` value
 - `%(alertsInterval)s` - expands to `interval` value
 
+## Making signals optional
+
+When defining signals from multiple sources, you can make some of the signals optional. In this case, rendering will not throw a validation error that signal is missing for the specific source, while internal 'stub' type will be used and empty panel will be rendered instead.
 
 ## Example 1: From JSON
 

--- a/common-lib/common/signal/signal.libsonnet
+++ b/common-lib/common/signal/signal.libsonnet
@@ -152,6 +152,7 @@ local stub = import './stub.libsonnet';
            else
              super.addSignal(
                name=std.get(signalsJson.signals[s], 'name', error 'Must provide name'),
+               nameShort=std.get(signalsJson.signals[s], 'nameShort', name),
                type='stub',
                description=std.get(signalsJson.signals[s], 'description', ''),
              ))

--- a/common-lib/common/signal/stub.libsonnet
+++ b/common-lib/common/signal/stub.libsonnet
@@ -1,13 +1,18 @@
 local g = import '../g.libsonnet';
 local utils = import '../utils.libsonnet';
 local signalUtils = import './utils.libsonnet';
-
 {
   new(
     signalName,
     type,
   ): {
+    local this = self,
 
+    withTopK(limit): this,
+    withOffset(offset): this,
+    withFilteringSelectorMixin(mixin): this,
+    withExprWrappersMixin(wrappers=[]): this,
+    asOverride(name=signalName, override='byQuery', format='time_series'): {},
     asTarget():: {},
     asTableTarget():: {},
 
@@ -31,9 +36,13 @@ local signalUtils = import './utils.libsonnet';
     asTable(name=signalName, format)::
       self.asTimeSeries(),
 
+    asStatusHistory(name=signalName)::
+      self.asTimeSeries(),
+
     //Return as timeSeriesPanel
     asGauge(name=signalName)::
       self.asTimeSeries(),
+
     asTableColumn(override, format):: {},
   },
 

--- a/common-lib/common/signal/test_stub.libsonnet
+++ b/common-lib/common/signal/test_stub.libsonnet
@@ -1,0 +1,60 @@
+local signal = import './signal.libsonnet';
+local test = import 'jsonnetunit/test.libsonnet';
+
+local m1 = signal.init(
+  filteringSelector=['job="abc"'],
+).addSignal(
+  name='Go version',
+  nameShort='Version',
+  type='stub',
+  aggLevel='instance',
+  description='Go version.',
+  sourceMaps=[
+    {
+      expr: 'go_info{%(queriesSelector)s}',
+      infoLabel: 'version',
+      legendCustomTemplate: null,
+      aggKeepLabels: [],
+    },
+  ]
+);
+
+{
+
+  asStat:
+    {
+      local raw = m1.asStat(),
+      testResult: test.suite({
+        testTStitle: {
+          actual: raw.title,
+          expect: '',  // empty title for stub panels
+        },
+        testType: {
+          actual: raw.type,
+          expect: 'text',
+        },
+        testTSversion: {
+          actual: raw.pluginVersion,
+          expect: 'v11.0.0',
+        },
+        testTSUid: {
+          actual: raw.datasource,
+          expect: {
+            uid: '-- Mixed --',
+            type: 'datasource',
+          },
+        },
+      }),
+    },
+  asTableTarget:
+    {
+      local raw = m1.asTableTarget(),
+      testResult: test.suite({
+        testTStitle: {
+          actual: raw,
+          expect: {},
+        },
+      }),
+    },
+
+}

